File: ee/engine/auditsink/configure.go
package auditsink

// Reading the sinks an installation has actually configured, and plugging them
// in.
//
// Not MIT. Covered by the Antifailure Enterprise License; see ee/LICENSE.md.
//
// The rule is the one AF_SECRET_SOURCES and AF_ORG_POLICY_FILE already follow,
// and it is sharper here than in either.
//
// Somebody who writes AF_AUDIT_SINKS=syslog has said that every privileged
// action this engine takes must be forwarded to their SIEM. Starting anyway
// with the sink unbuilt means every action goes unforwarded and nothing in the
// output says so, which is indistinguishable from a quiet week. That is not a
// degraded feature, it is a compliance control that reports itself as held
// while holding nothing, and it is the exact defect this whole lane exists to
// close. So a named sink that cannot be built stops the process with the reason.
//
// An unset variable registers nothing and prints nothing, which is the ordinary
// case. Nothing is ever auto-detected: a machine that happens to carry AWS
// credentials for something unrelated must not have this tool decide on its own
// to start writing that organization's audit trail into their bucket.

import (
	"context"
	"fmt"
	"os"
	"sort"
	"strings"

	"github.com/antifailure/antifailure/engine/pkg/extension"
)

// SinksEnv names the variable that lists which sinks to use.
const SinksEnv = "AF_AUDIT_SINKS"

// The variables each sink reads. Named as constants rather than written inline
// so that the error message listing what is missing and the code reading it
// cannot drift apart.
const (
	SyslogAddressEnv  = "AF_AUDIT_SYSLOG_ADDRESS"
	SyslogCAEnv       = "AF_AUDIT_SYSLOG_CA_FILE"
	SyslogCertEnv     = "AF_AUDIT_SYSLOG_CERT_FILE"
	SyslogKeyEnv      = "AF_AUDIT_SYSLOG_KEY_FILE"
	SyslogHostnameEnv = "AF_AUDIT_SYSLOG_HOSTNAME"

	WebhookURLEnv        = "AF_AUDIT_WEBHOOK_URL"
	WebhookSecretEnv     = "AF_AUDIT_WEBHOOK_SECRET"
	WebhookHeaderEnv     = "AF_AUDIT_WEBHOOK_HEADER"
	WebhookDeadLetterEnv = "AF_AUDIT_WEBHOOK_DEAD_LETTER_FILE"

	ObjectStoreURLEnv = "AF_AUDIT_OBJECT_STORE_URL"
)

// Known are the sinks this build can write to, for the message that lists them
// when somebody names one that does not exist.
func Known() []string {
	out := []string{"syslog", "webhook", "object_store"}
	sort.Strings(out)
	return out
}

// Sink is what this package registers: the interface, plus the name, so a
// caller can print what was configured without knowing which concrete type it
// got.
type Sink = extension.AuditSink

// FromEnvironment builds the sinks named by AF_AUDIT_SINKS, in order.
//
// The order is the order given and it is the order they are written in, so an
// installation that wants its SIEM to see an entry before its archive does can
// say so. Every sink is written to regardless of whether an earlier one failed:
// see Registry.Audit, which collects failures rather than stopping, because a
// SIEM outage must not cost the archive its copy.
func FromEnvironment(getenv func(string) string) ([]Sink, error) {
	if getenv == nil {
		getenv = os.Getenv
	}
	names := strings.FieldsFunc(getenv(SinksEnv), func(r rune) bool {
		return r == ',' || r == ' '
	})
	if len(names) == 0 {
		return nil, nil
	}

	out := make([]Sink, 0, len(names))
	for _, raw := range names {
		name := strings.ToLower(strings.TrimSpace(raw))
		sink, err := build(name, getenv)
		if err != nil {
			return nil, fmt.Errorf("%s names %s and it cannot be used: %w", SinksEnv, name, err)
		}
		out = append(out, sink)
	}
	return out, nil
}

func build(name string, getenv func(string) string) (Sink, error) {
	switch name {
	case "syslog":
		return NewSyslog(SyslogConfig{
			Address:  getenv(SyslogAddressEnv),
			CAFile:   getenv(SyslogCAEnv),
			CertFile: getenv(SyslogCertEnv),
			KeyFile:  getenv(SyslogKeyEnv),
			Hostname: getenv(SyslogHostnameEnv),
		})
	case "webhook":
		return NewWebhook(WebhookConfig{
			URL:            getenv(WebhookURLEnv),
			Secret:         getenv(WebhookSecretEnv),
			Header:         getenv(WebhookHeaderEnv),
			DeadLetterFile: getenv(WebhookDeadLetterEnv),
		})
	case "object_store", "objectstore", "s3", "azure_blob":
		// Four spellings for one sink, because the URL is what decides which
		// protocol is spoken and somebody who writes s3 means this. Refusing
		// the obvious name and naming the canonical one in the error would be a
		// round trip for nothing.
		return NewObjectStore(ObjectStoreConfig{
			URL:    getenv(ObjectStoreURLEnv),
			Getenv: getenv,
		})
	default:
		return nil, fmt.Errorf("there is no such sink in this build; it knows %s",
			strings.Join(Known(), ", "))
	}
}

// RegisterFromEnvironment reads the configured sinks and plugs them in.
//
// The one call an embedding binary makes. It returns the sinks so the binary
// can print which destinations are in force, and so that a binary that forgot
// to print them still cannot forget to register them: there is no way to get a
// sink out of here without it already being in the registry. That property is
// what was missing before this lane, in exactly this shape: policyenforce.Hook
// was written, tested and never constructed by any binary, and audit_stream had
// no implementation to construct at all.
func RegisterFromEnvironment(reg *extension.Registry, getenv func(string) string) ([]Sink, error) {
	sinks, err := FromEnvironment(getenv)
	if err != nil {
		return nil, err
	}
	for _, s := range sinks {
		reg.AddAuditSink(s)
	}
	return sinks, nil
}

// Describe names each configured sink, for a startup line and for af doctor.
//
// Worth printing for the same reason the secret sources are: an operator asking
// why an action did not appear in their SIEM needs to know whether this
// installation was forwarding at all before they go looking at the receiver.
// Each sink's Name is already written to carry no credential.
func Describe(sinks []Sink) []string {
	out := make([]string, 0, len(sinks))
	for _, s := range sinks {
		out = append(out, s.Name())
	}
	return out
}

// Unlicensed reports whether entries will actually be forwarded.
//
// Configured and licensed are two different things and an operator needs to be
// able to tell them apart. A sink that is configured under a lapsed licence
// accepts every entry and writes none, which is the correct behaviour and is
// also silence, so the binary says so once at startup rather than leaving
// somebody to discover it from an empty dashboard.
func Unlicensed(ctx context.Context, sinks []Sink) bool {
	return len(sinks) > 0 && !permitted(ctx)
}

File: ee/web/audit/src/configure.ts
// Reading the sink an installation has actually configured.
//
// Not MIT. Covered by the Antifailure Enterprise License; see ee/LICENSE.md.
//
// The rule is `ee/engine/auditsink/configure.go`'s and it is the same rule for
// the same reason. Somebody who writes AF_AUDIT_STREAM_SINK=splunk has said
// that every privileged action in this control plane must reach their SIEM.
// Starting anyway with the sink unbuilt means nothing is forwarded and nothing
// says so, which is indistinguishable from a quiet week. That is not a degraded
// feature; it is a compliance control reporting itself as held while holding
// nothing, which is the exact defect this whole file exists downstream of.
//
// So a named sink that cannot be built REFUSES, with the variable that is
// missing named in the message. An unset variable builds nothing and says so
// once, which is the ordinary case. Nothing is ever detected: a control plane
// that happens to carry a webhook URL for something unrelated must not decide
// on its own to start posting an organization's audit trail to it.
//
// WHAT IS NOT HERE, SAID PLAINLY RATHER THAN IMPLIED BY ITS ABSENCE.
// `ObjectStoreSink` exists in sinks.ts and cannot be built from the
// environment, because it takes a `put` callback and this side of the product
// has no S3 or Blob signer to supply one. It is reachable by an embedder that
// passes a Sink directly and it is not reachable from a container's
// configuration, and `Known()` below lists only what an operator can actually
// turn on. A name accepted here that then wrote nowhere would be the same
// failure one layer along.

import {
  EventHubsSink,
  SplunkSink,
  WebhookSink,
  type Fetcher,
} from './sinks.ts'
import type { Sink } from './sink.ts'

/** Names the variable that chooses the destination. */
export const SinkEnv = 'AF_AUDIT_STREAM_SINK'

/** The variables each sink reads. Constants rather than inline strings so that
 *  the message listing what is missing and the code reading it cannot drift. */
export const KeyEnv = 'AF_AUDIT_STREAM_KEY'
export const IntervalEnv = 'AF_AUDIT_STREAM_INTERVAL_MS'
export const BatchEnv = 'AF_AUDIT_STREAM_BATCH'
export const DeliveryBatchEnv = 'AF_AUDIT_STREAM_DELIVERY_BATCH'

export const SplunkUrlEnv = 'AF_AUDIT_STREAM_SPLUNK_URL'
export const SplunkTokenEnv = 'AF_AUDIT_STREAM_SPLUNK_TOKEN'
export const SplunkIndexEnv = 'AF_AUDIT_STREAM_SPLUNK_INDEX'
export const SplunkSourcetypeEnv = 'AF_AUDIT_STREAM_SPLUNK_SOURCETYPE'

export const EventHubsUrlEnv = 'AF_AUDIT_STREAM_EVENT_HUBS_URL'
export const EventHubsAuthorizationEnv = 'AF_AUDIT_STREAM_EVENT_HUBS_AUTHORIZATION'

export const WebhookUrlEnv = 'AF_AUDIT_STREAM_WEBHOOK_URL'
export const WebhookSecretEnv = 'AF_AUDIT_STREAM_WEBHOOK_SECRET'

/** The destinations an operator can name. Sorted, for the message that lists
 *  them when somebody names one that does not exist. */
export function known(): string[] {
  return ['event_hubs', 'splunk', 'webhook']
}

/** A configuration mistake, separated from every other error so the caller can
 *  refuse to start rather than log and continue. */
export class SinkRefused extends Error {}

export interface StreamConfig {
  sink: Sink
  key: string
  intervalMs: number
  /** Entries one pass reads. */
  batchSize: number
  /** Entries one delivery carries. A different question from the one above: a
   *  pass size is how fast the forwarder catches up and a delivery size is what
   *  one request to somebody else's collector may hold. Defaults to the pass
   *  size, so an operator who does not care sets one number. */
  deliveryBatchSize: number
}

/** How often a pass runs when nothing says otherwise. Ten seconds, because an
 *  audit stream is read during an incident and five minutes of lag is five
 *  minutes of a security team seeing nothing after something started. */
export const DEFAULT_INTERVAL_MS = 10_000

/** How many entries one pass reads when nothing says otherwise. */
export const DEFAULT_BATCH = 500

/**
 * Builds the configured sink, or null when none is configured.
 *
 * `fetch` is a parameter for the reason sinks.ts gives about its own: it is
 * what lets an operator route through their own proxy or add a header their
 * gateway requires, and it is what lets a test hold a real receiver.
 */
export function fromEnvironment(
  env: NodeJS.ProcessEnv,
  fetcher: Fetcher = (url, init) => fetch(url, init),
): StreamConfig | null {
  const name = (env[SinkEnv] ?? '').trim().toLowerCase()
  if (name === '') return null

  if (!known().includes(name)) {
    throw new SinkRefused(
      `${SinkEnv} names ${name} and this build has no such sink. It knows ${known().join(', ')}. ` +
        'An object store sink exists in the code and cannot be configured from the environment, ' +
        'because it needs a signer this side of the product does not carry.',
    )
  }

  const key = (env[KeyEnv] ?? '').trim()
  if (key === '') {
    // Required rather than defaulted. The manifest is what lets an auditor
    // check a batch without asking this control plane anything, and a manifest
    // signed under a key everybody knows is a decoration rather than evidence.
    throw new SinkRefused(
      `${SinkEnv} is set to ${name} and ${KeyEnv} is not. Batch manifests are signed under it, ` +
        'and a signature under a key nobody chose proves nothing. Generate one with ' +
        '`openssl rand -base64 32`.',
    )
  }

  return { sink: build(name, env, fetcher), key, ...scheduleFromEnvironment(env) }
}

export interface Schedule {
  intervalMs: number
  batchSize: number
  deliveryBatchSize: number
}

/**
 * How often and how much, whether or not the installation has a sink of its own.
 *
 * Separate from `fromEnvironment` because an installation whose organizations
 * choose their own destinations runs a forwarder with no installation sink at
 * all, and the operator's interval and batch settings must still apply to it.
 * Refuses a malformed number for the reason `positive` gives.
 */
export function scheduleFromEnvironment(env: NodeJS.ProcessEnv): Schedule {
  const batchSize = positive(env, BatchEnv, DEFAULT_BATCH)
  return {
    intervalMs: positive(env, IntervalEnv, DEFAULT_INTERVAL_MS),
    batchSize,
    deliveryBatchSize: positive(env, DeliveryBatchEnv, batchSize),
  }
}

function build(name: string, env: NodeJS.ProcessEnv, fetcher: Fetcher): Sink {
  switch (name) {
    case 'splunk':
      return new SplunkSink({
        url: required(env, SplunkUrlEnv, name),
        token: required(env, SplunkTokenEnv, name),
        fetch: fetcher,
        ...(env[SplunkIndexEnv] ? { index: env[SplunkIndexEnv] } : {}),
        ...(env[SplunkSourcetypeEnv] ? { sourcetype: env[SplunkSourcetypeEnv] } : {}),
      })
    case 'event_hubs':
      return new EventHubsSink({
        url: required(env, EventHubsUrlEnv, name),
        authorization: required(env, EventHubsAuthorizationEnv, name),
        fetch: fetcher,
      })
    default:
      return new WebhookSink({
        url: required(env, WebhookUrlEnv, name),
        secret: required(env, WebhookSecretEnv, name),
        fetch: fetcher,
      })
  }
}

function required(env: NodeJS.ProcessEnv, variable: string, sink: string): string {
  const value = (env[variable] ?? '').trim()
  if (value === '') {
    throw new SinkRefused(
      `${SinkEnv} is set to ${sink} and ${variable} is not set. Refusing to start rather than ` +
        'forwarding nothing quietly.',
    )
  }
  return value
}

/** A whole number above zero, or the default. Refuses anything else rather than
 *  falling back, because a typo that silently becomes ten seconds is a
 *  configuration an operator believes they changed. */
function positive(env: NodeJS.ProcessEnv, variable: string, fallback: number): number {
  const raw = (env[variable] ?? '').trim()
  if (raw === '') return fallback
  const value = Number(raw)
  if (!Number.isInteger(value) || value <= 0) {
    throw new SinkRefused(
      `${variable} is ${JSON.stringify(raw)}; it has to be a whole number above zero.`,
    )
  }
  return value
}
